enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,423 @@
+- Provide migration of certificates from current solution.
+- Provide management of previously added certificates on the Route.
+
+## Proposal
+
+This enhancement proposes introducing a new controller (secret-injector) in [route-controller-manager](https://github.com/openshift/route-controller-manager)
+to manage a new annotation on the Route object. This annotation enables the users to
+provide a reference to a Secret containing the serving cert/key pair that will
+be injected to `.spec.tls` and will be served by OpenShift router. This annotation
+will be given a higher preference if route CR also has `.spec.tls.certificate`
+and `.spec.tls.key` fields set.
+
+### Workflow Description
+
+End users will have 2 possible variations when using this functionality,
+
+- Create a route for the user workload and this is completely managed by the end user.
+- Create an ingress for the user workload and a managed route is created automatically
+  by the ingress-to-route controller.
+
+Both these workflows will support integrating with third party certificate management
+solutions like cert-manager with the annotation described below.
+
+**When a Route is directly used to expose user workload**
+
+- The end user must have generated the serving certificate as a pre requisite
+  using third-party systems like cert-manager.
+- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CR must be created in the same namespace
+  where the Route is going to be created.
+- To expose a user workload, the user would create a new Route with an annotation
+  referencing the generated secret that was created in the previous step.
+- If the secret that is referenced exists and has a successfully generated
+  cert/key pair generated, the secret-injector controller copies the certificate data
+  into the route at `.spec.tls.certificate` and `.spec.tls.key`.
+
+**When an Ingress is used to expose user workload**
+
+- The end user must have generated the serving certificate as a pre requisite
+  using third-party systems like cert-manager.
+- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CR must be created in the same namespace
+  where the Ingress is going to be created.
+- To expose a user workload, he would need to create a new Ingress with the generated secret
+  referenced in `.spec.tls.secretName`.
+- If the secret CR that is referenced exists and has a successfully generated
+  cert/key pair, the ingress-to-route controller copies this certificate data
+  into the route at `.spec.tls.certificate` and `.spec.tls.key`.
+
+_Note_: The same workflow would apply to updating an existing Route to integrate
+with cert-manager. The new certificate data will overwrite the `.spec.tls.certificate`
+and `.spec.tls.key` in the route CR since the annotation gets a higher preference.
+
+#### Certificate Life cycle
```

Comment:
It'd be good to detail the behavioral variations when ingress is configured with a secret name (not managed by cert-manager), route has annotation for secret.